packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -568,18 +670,73 @@ function updateHostText(current, workInProgress) {
   return null;
 }
 
+function resolveDefaultProps(Component, baseProps) {
+  if (Component && Component.defaultProps) {
+    // Resolve default props. Taken from ReactElement
+    const props = Object.assign({}, baseProps);
+    const defaultProps = Component.defaultProps;
+    for (let propName in defaultProps) {
+      if (props[propName] === undefined) {
+        props[propName] = defaultProps[propName];
+      }
+    }
+    return props;
+  }
+  return baseProps;
+}
+
 function mountIndeterminateComponent(
   current,
   workInProgress,
+  Component,
   renderExpirationTime,
 ) {
   invariant(
     current === null,
     'An indeterminate component should never have mounted. This error is ' +
       'likely caused by a bug in React. Please file an issue.',
   );
-  const fn = workInProgress.type;
+
   const props = workInProgress.pendingProps;
+  if (
+    Component !== null &&
+    Component !== undefined &&
+    typeof Component.then === 'function'
+  ) {
+    Component = readLazyComponentType(Component);
+    if (typeof Component === 'function') {
+      if (shouldConstruct(Component)) {
+        workInProgress.tag = ClassComponentLazy;
+        return updateClassComponentLazy(
+          current,
+          workInProgress,
+          Component,
+          renderExpirationTime,
+        );
+      } else {
+        workInProgress.tag = FunctionalComponentLazy;
+        return updateFunctionalComponentLazy(
+          current,
+          workInProgress,
+          Component,
+          renderExpirationTime,
+        );
+      }
+    } else if (
+      Component !== undefined &&
+      Component !== null &&
+      Component.$$typeof
+    ) {
+      workInProgress.tag = ForwardRefLazy;
+      return updateForwardRefLazy(
+        current,
+        workInProgress,
+        Component,
+        renderExpirationTime,
+      );
+    }
+  }
```

Comment:
Does this mean that we actually want to keep the concept of IndeterminateComponent even if we deprecate the "module pattern" components?

Reply:
Seems like it